proposals/p1885.md

```diff
@@ -175,111 +195,176 @@ https://stackoverflow.com/questions/35810674/in-go-is-it-possible-to-iterate-ove
 Provide interfaces that can be implemented by user types to enable support for
 ranged-for loops.
 
-We propose three interfaces:
-
--   [**Indexable**](#indexable): Implements index-based random access
--   [**Iterable**](#iterable): Implements iterator-based sequential access
--   [**Consumable**](#consumable): Implements single-use sequential access
-
-Users would benefit from providing an `external impl`, to limit class members
-pollution.
+We propose a new [**Iterable**](#iterable) interface, for iterator-based
+sequential access, and usage with range-based `for`.
 
 ## Details
 
-This proposal covers support for 2 different type of containers:
-
--   Random access:
-    -   `Indexable`
--   Sequential access:
-    -   `Iterable`, `Iterator`
-    -   `Consumable`
-
-Due to the semantic and functional difference between these interfaces, they
-should not conflict with the [One way](/docs/project/principles/one_way.md).
+This proposal exposes an Iterable-type interface, relying on 2 new interfaces:
+`Iterable` and `Iterator`.
 
-In absence of strong requirements from the algorithm, this would require future
-algorithms to:
+A user implements these interfaces to enable range-based `for` support. Carbon
+would make use of the `Iterable` to retrieve an `Iterator` and iterate until
+completed or interrupted.
 
--   A. explicitly support multiple interfaces: expressive, and specialize
-    implementations based on type features (likely preferred)
--   or B. define adapters or implicit conversions to respectively simplify
-    algorithms implementation or end-user usage
+### Iterable
 
-### Indexable
+The `Iterable` interface represents a generic sequential-access type of
+container. It relies on two new interfaces:
 
-`Indexable` represents a container that supports index-based random access.
+-   `Iterable` and `MutableIterable`: Interface for a type that supports being
+    iterated over
+-   `Iterator`: Interface for an iterator-like object
 
 #### Range-based `for` behavior
 
--   Get container size (`size`)
--   Iterate from `[0;size[`, returning current item at each iteration
+-   Get an `Iterator` type with `GetIterator()` or `GetMutableIterator()`
+-   Call `Advance() -> bool`
+    -   Break the loop if the result is `false`
+-   Call `Get()` and return the result
 
 #### Interface and usage
 
-```rust
-// Carbon definition
-interface Indexable() {
-    let ElementType:! Type
-    fn GetSize[me: Self]() -> u32;
-    fn GetAt[me: Self](pos: u32) -> ElementType;
+These interfaces are defined as:
+
+```carbon
+// Basic immutable iterator
+interface Iterator {
```

Comment:
Why go with the iterator approach instead of the cursor approach? I think the cursor approach has a number of advantages, and is what I've heard @chandlerc prefers. For example, I think the cursor approach supports iterating through r-value arrays.

Reply:
I'm sorry, it was not my intention to make an argument from authority. I was trying to say something about how I had heard different stories in different venues from different people and I was hoping they would connect to discuss it. My understanding of the goals of the proposal process is to gather and represent the different opinions and approaches, ideally before the proposal is sent to the leads for final ratification and I perceived a gap. Unfortunately, my attempt to add @chandlerc to the conversation did not lead to that outcome. Next time I will try harder to make it clear that this was an invitation to continue a conversion rather than attempt to end it.

Reply:
Moved to cursor approach following discussion with leads.